### PR TITLE
Random op

### PR DESCRIPTION
**Pull request by dzhwinter**
there is two part need to be discussed. 
One is the `dynamc_cast` part of implementing polynomial.  Another one is removing the const qualify key in `Op.Run` interface.

### CODE REVIEW

File: paddle/framework/operator.h [hunk 1]
```diff
@@ -88,7 +88,7 @@ class OperatorBase {
 
   /// Net will call this function to Run an op.
   virtual void Run(const std::shared_ptr<Scope>& scope,
-                   const platform::DeviceContext& dev_ctx) const = 0;
+                   platform::DeviceContext& dev_ctx) const = 0;
```
Comment (dzhwinter) [inline paddle/framework/operator.h]: 由于cuda的randGenerator和stream绑定，stream只存在于CUDADeviceContext里，只能在Op设置种子，生成Generator，违反了Op Run的时候不能修改context。因此去掉了const 约束。

尝试过解决办法，全局一个static generator。random_seed可以只在创建CUDADeviceContext一次指定，但是不能创建一个全局的Generator，否则随机性有问题。

File: paddle/operators/random_op.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+public:
+  void Compute(const framework::KernelContext& context) const override {
+    auto mean = context.op_.GetAttr<T>("mean");
+    auto std = context.op_.GetAttr<T>("std");
+    auto seed = context.op_.GetAttr<T>("seed");
+    auto* output = context.Output(0)->GetMutable<framework::Tensor>();
+    auto place = context.GetPlace();
+    if (platform::is_cpu_place(place)) {
+      Gaussian(
+          dynamic_cast<platform::CPUDeviceContext*>(context.device_context_),
```
Comment (Canpio) [inline paddle/operators/random_op.h]: 这里用`dynamic_cast`效率可能比较低下，应该不是最佳的实现方式。
导致这里使用`dynamic_cast`的原因是`KernalContext`里用一个父类`DeviceContext`的引用去保存子类`CPUDeviceContext`和`CUDADeviceContext`:
```cpp
const platform::DeviceContext& device_context_;
```
导致子类具体的类型信息丢失，在这里只能显示用`dynamic_cast`转换。

所以，是否可以考虑在`KernelContext`中用
```cpp
boost::variant<CPUDeviceContex, CUDADeviceContex>
```
来保存device context信息，这样就可以比较容易的知道具体的类型信息。

File: paddle/operators/random_op.cc [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    PADDLE_ENFORCE(inputs.size() == 0, "Input size of RandomOp must be zero.");
+    PADDLE_ENFORCE(outputs.size() == 1, "Output size of RandomOp must be one.");
+    PADDLE_ENFORCE(outputs[0] != nullptr,
+                   "Outputs of RandomOp must all be set.");
+    outputs[0]->Resize(
+        framework::make_ddim(this->GetAttr<std::vector<int>>("shape")));
+  }
+};
+
+class RandomOpMaker : public framework::OpProtoAndCheckerMaker {
```
Comment (emailweixu) [inline paddle/operators/random_op.cc]: "random" is too general. There can be many different types of random. We should use a more specific name such as "GaussionRandom"

File: paddle/framework/operator.h (hunk 1 above)
Comment (QiJune) [inline paddle/framework/operator.h]: 如果seed固定的话，生成的序列也是一样的。
目前paddle/caffe2的随机数生成也都是提前设定好一个seed，后续不会改变。请教一下@hedaoyuan，不知道这样是否有问题。

File: paddle/operators/random_op.h (hunk 2 above)
Comment (QiJune) [inline paddle/operators/random_op.h]: Place和DeviceContext的类型是可以一一对应的。可以在编译期根据Place的类型确定DeviceContext的类型。
在.h里面声明
```
template <typename Place, typename T>
class RandomOpKernel : public framework::OpKernel {
public:
  void Compute(const framework::KernelContext& context) const override;  
}
```

分别在.cc和.cu里面偏特化
```
template <typename T>
class RandomOpKernel<CPUPlace, T> : public framework::OpKernel {
  void Compute() {
    CPUDeviceContext* dc = static_cast<CPUDeviceContext*>(KernelContext.device_context_);
  }
}
```

.cu
```
template <typename T>
class RandomOpKernel<GPUPlace, T> : public framework::OpKernel {
  void Compute() {
    CUDADeviceContext* dc = static_cast<CUDADeviceContext*>(KernelContext.device_context_);
  }
}
```

File: paddle/framework/operator.h (hunk 1 above)
Comment (hedaoyuan) [inline paddle/framework/operator.h]: 每个线程的seed只需要设置一次就行了。

File: paddle/operators/random_op.cc (hunk 3 above)
Comment (dzhwinter) [inline paddle/operators/random_op.cc]: Done.

File: paddle/operators/random_op.h (hunk 2 above)
Comment (dzhwinter) [inline paddle/operators/random_op.h]: Thanks for your comments! The template specialization can fix it elegantly. fix Done.

File: paddle/framework/operator.h (hunk 1 above)
Comment (dzhwinter) [inline paddle/framework/operator.h]: fix done.

Comment (dzhwinter) [inline paddle/framework/operator.h]: const DeviceContext是否必要？ `cublas_handle` 等函数需要运行期间修改context

File: paddle/operators/gaussian_random_op.cc [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  void Compute(const framework::KernelContext& context) const override {
+    auto mean = context.op_.GetAttr<T>("mean");
+    auto std = context.op_.GetAttr<T>("std");
+    auto* output = context.Output(0)->GetMutable<framework::Tensor>();
+    T* r = output->mutable_data<T>(context.GetPlace());
+    auto ctx =
+        static_cast<const platform::CPUDeviceContext*>(context.device_context_);
+    // generator need to modify context
+    auto g = const_cast<platform::CPUDeviceContext*>(ctx)->RandGenerator();
+    std::normal_distribution<T> distribution(mean, std);
```
Comment (gangliao) [inline paddle/operators/gaussian_random_op.cc]: Only `std normal distribution` in here. Did you also consider other distribution soon? like `uniform distribution`.

```
std::uniform_real_distribution<T> uniform(0.0, 1.0);
```

File: paddle/platform/device_context.h [hunk 5]
```diff
@@ -40,7 +41,10 @@ class DeviceContext {
 class CPUDeviceContext : public DeviceContext {
  public:
   typedef std::mt19937 random_generator_type;
-  CPUDeviceContext() { eigen_device_.reset(new Eigen::DefaultDevice()); }
+  CPUDeviceContext() {
+    random_seed_ = std::chrono::system_clock::now().time_since_epoch().count();
```
Comment (Canpio) [inline paddle/platform/device_context.h]: So now we are not able to manually specify random seed?

File: paddle/operators/gaussian_random_op.cc (hunk 4 above)
Comment (dzhwinter) [inline paddle/operators/gaussian_random_op.cc]: @reyoung implement the uniform random generator in thrust.  
see #3293 for detail.

File: paddle/platform/device_context.h (hunk 5 above)
Comment (dzhwinter) [inline paddle/platform/device_context.h]: deleted the seeding function in context. 
 it is the place that we need to modify the Execution context. currently, move the set seed to random op. We need a global configuration to unify the random seed in the future. Maybe we can fix it in next PR.

File: paddle/operators/gaussian_random_op.cc [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+namespace paddle {
+namespace operators {
+
+template <typename T>
+class GaussianRandomKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    T mean = static_cast<T>(context.op_.GetAttr<T>("mean"));
+    T std = static_cast<T>(context.op_.GetAttr<T>("std"));
```
Comment (Canpio) [inline paddle/operators/gaussian_random_op.cc]: Do we need to add some `assert` to make sure `T` is `int` or `float`?

File: paddle/operators/gaussian_random_op.cc [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+    // TODO(dzh): attribute does not support unsigned int.
+    // And we need a global random seed configuration.
+    int seed = context.op_.GetAttr<int>("seed");
+    if (seed == 0) {
+      seed = std::random_device()();
+    }
+    std::mt19937 g(seed);
+    std::normal_distribution<T> distribution(mean, std);
+    for (int i = 0; i < framework::product(tensor->dims()); ++i) {
```
Comment (Canpio) [inline paddle/operators/gaussian_random_op.cc]: Can we accelerate the code by calculating `framework::product()` before the loop? I'm not sure whether the compiler will do the optimization for us.

File: paddle/operators/gaussian_random_op.cu [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace paddle {
+namespace operators {
+
+template <typename T>
+class GaussianRandomKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    T mean = static_cast<T>(context.op_.GetAttr<T>("mean"));
+    T std = static_cast<T>(context.op_.GetAttr<T>("std"));
```
Comment (Canpio) [inline paddle/operators/gaussian_random_op.cu]: The same question in https://github.com/PaddlePaddle/Paddle/pull/3060/files#diff-86aabb5beb3f333cf191423101025d1fR26

File: paddle/operators/gaussian_random_op.cc (hunk 6 above)
Comment (dzhwinter) [inline paddle/operators/gaussian_random_op.cc]: just unified style with the uniform random operator.  Revert these things to old code.

File: paddle/operators/gaussian_random_op.cc (hunk 7 above)
Comment (dzhwinter) [inline paddle/operators/gaussian_random_op.cc]: the compiler can't be optimized at all. tensor shape only can be referenced in runtime. Done.

File: paddle/operators/gaussian_random_op.cu (hunk 8 above)
Comment (dzhwinter) [inline paddle/operators/gaussian_random_op.cu]: Done.

**Pull request merged**